### PR TITLE
Updating Makefile/.drone.yml to reflect current patterns in other plugins

### PR DESCRIPTION
**Pull request by gtaylor**
I ran into some issues when trying to build Docker images with the old Makefile:
```
$ docker run --rm -it plugins/drone-google-cloudstorage
no such file or directory
Error response from daemon: Cannot start container 616d22fc9506e847d990c076c1ebb232ad2f4598a9d9b91204e0f1b5db7d9a15: [8] System error: no such file or directory
```
The binary was present, but it looked like something was wrong with the linkage. I couldn't figure out specifically what was wrong, but noticed that I had no problem building most of the other more recent plugins. I compared Makefiles between the other plugins that were building+running correctly, only to find a different set of options being specified.

After updating drone-google-cloudstorage to reflect these new build params (the ldflags in particular), the builds started working.

**This PR updates .drone.yml, so the hash will change.**

### CODE REVIEW

**Comment by gtaylor** [comment]
@bradrydzewski This one will require your involvement, due to the .drone.yml checksum change.

**Comment by bradrydzewski** [comment]
LGTM

**Comment by Bugagazavr** [comment]
LGTM

**Comment by gtaylor** [comment]
Got the two LGTMs, can someone merge this?

File: Makefile [hunk 1]
```diff
@@ -1,8 +1,25 @@
...
+clean:
+	go clean -i ./...
+
+deps:
+	go get -t ./...
+
+test:
+	go test -cover ./...
+
+build:
```
Comment (crhym3) [inline Makefile]: why change this to `build`? make won't be able to identify what it's actually building anymore.

another issue with this is there are no target prerequisites so `make build` will always execute unconditionally.

File: Makefile [hunk 2]
```diff
@@ -1,8 +1,25 @@
...
+export GOOS ?= linux
+export GOARCH ?= amd64
+export CGO_ENABLED ?= 0
 
-$(BIN): $(wildcard *.go)
-	GOOS=linux GOARCH=amd64 go build
+CI_BUILD_NUMBER ?= 0
+
+LDFLAGS += -X "main.buildDate=$(shell date -u '+%Y-%m-%d %H:%M:%S %Z')"
+LDFLAGS += -X "main.build=$(CI_BUILD_NUMBER)"
```
Comment (crhym3) [inline Makefile]: The resulting binary doesn't currently seem to declare `build` nor `buildDate`. This flag option won't have any effect.

File: Makefile (hunk 1 above)
Comment (tboerger) [inline Makefile]: This is simply the Makefile we have added to various plugins.

File: Makefile [hunk 3]
```diff
@@ -1,8 +1,25 @@
...
+	go get -t ./...
+
+test:
+	go test -cover ./...
+
+build:
+	go build -ldflags '-s -w $(LDFLAGS)'
+
+docker:
+	docker build --rm=true -t plugins/drone-google-cloudstorage .
```
Comment (crhym3) [inline Makefile]: A better way to do his (esp. given the defined `.PHONY` which suggest a certain usage pattern) is something like:

```
docker: .docker
.docker: drone-google-cloudstorage Dockerfile
	docker build --rm=true -t plugins/drone-google-cloudstorage .
	touch .docker
```

That way the image will be built only when source code or Dockerfile have changed since the last build.

File: Makefile (hunk 2 above)
Comment (tboerger) [inline Makefile]: That should be added to get some information in the build log to see if it's really a up2date version of the plugin.

File: .drone.yml [hunk 4]
```diff
@@ -2,13 +2,10 @@ build:
   image: golang:1.5
   environment:
     - GO15VENDOREXPERIMENT=1
```
Comment (tboerger) [inline .drone.yml]: Drop this as well, we are not using any vendoring within the plugins yet.

File: Makefile (hunk 2 above)
Comment (crhym3) [inline Makefile]: Exactly my point. How are you going to get that info at the moment? It would make sense to add related/dependent changes in one PR as opposed to scattered across unrelated changes.

Comment (tboerger) [inline Makefile]: Those flags are added to various other plugins as well.

Comment (tboerger) [inline Makefile]: > Exactly my point. How are you going to get that info at the moment? It would make sense to add related/dependent changes in one PR as opposed to scattered across unrelated changes.

Sure, this is something @gtaylor can still add to the first line of `main` function.

Comment (crhym3) [inline Makefile]: Even if he did, what are you gonna do then, run `readelf` or `nm` on the binary and look up `buildDate` symbol? :)

File: Makefile (hunk 1 above)
Comment (crhym3) [inline Makefile]: See my comments on the `docker` target below to explain the point I'm making.

File: Makefile (hunk 2 above)
Comment (crhym3) [inline Makefile]: Even then I bet you'll never find it until it's actually used somewhere in the code.

**Comment by crhym3** [comment]
Sorry but given this PR description, the change doesn't make sense: `LDFLAGS` currently add absolutely nothing new. There's no `build` nor `buildDate` declared in the binary.

**Comment by crhym3** [comment]
What's the connection between `docker run --rm -it plugins/drone-google-cloudstorage`, which pulls the image from docker hub, and local Makefile?

**Comment by gtaylor** [comment]
I don't have strong feelings on any of this, and am not familiar enough with the Go linker/compiler to have much to offer. The big thing I'm interested in is that once I copy/pasted this from one of the other plugins (most seem to be using variations of this Makefile now), the issue was resolved. 

Not at all scientific, but since it seemed like we were using this Makefile elsewhere, I figured if it fixed the issue and made us more consistent it is a net win.

I'd be glad to close this PR if you'd like to propose an alternative Makefile, @crhym3. I'm mostly just wanting to stop having to run a custom build of this plugin to use it after making some slight changes to the way auth is done.

**Comment by crhym3** [comment]
I don't think I understand what the actual issue is here.

Could you post the output of `make docker && docker run --rm -it plugins/drone-google-cloudstorage` running from HEAD of this repo?

**Comment by tboerger** [comment]
> Could you post the output of make docker && docker run --rm -it plugins/drone-google-cloudstorage running from HEAD of this repo?

On a Mac with docker-machine it seems to be fine on master.

```
# make docker
GOOS=linux GOARCH=amd64 go build
docker build --rm -t plugins/drone-google-cloudstorage .
Sending build context to Docker daemon 11.55 MB
Step 1 : FROM alpine
latest: Pulling from library/alpine

Digest: sha256:78a756d480bcbc35db6dcc05b08228a39b32c2b2c7e02336a2dcaa196547a41d
Status: Downloaded newer image for alpine:latest
 ---> 74e49af2062e
Step 2 : RUN apk add --update ca-certificates && rm -rf /var/cache/apk/*
 ---> Running in fb757856eedb
fetch http://dl-4.alpinelinux.org/alpine/v3.3/main/x86_64/APKINDEX.tar.gz
fetch http://dl-4.alpinelinux.org/alpine/v3.3/community/x86_64/APKINDEX.tar.gz
(1/2) Installing openssl (1.0.2e-r0)
(2/2) Installing ca-certificates (20160104-r2)
Executing busybox-1.24.1-r7.trigger
Executing ca-certificates-20160104-r2.trigger
OK: 6 MiB in 13 packages
 ---> 5e284b562d46
Removing intermediate container fb757856eedb
Step 3 : ADD drone-google-cloudstorage /bin/
 ---> 02f358a55159
Removing intermediate container b1b5ffcf8180
Step 4 : ENTRYPOINT /bin/drone-google-cloudstorage
 ---> Running in fae6b2c34b37
 ---> 16e9aeaf516f
Removing intermediate container fae6b2c34b37
Successfully built 16e9aeaf516f
```

```
# docker run --rm -i plugins/drone-google-cloudstorage <<EOF
> {
>     "repo": {
>         "clone_url": "git://github.com/drone/drone",
>         "full_name": "drone/drone"
>     },
>     "build": {
>         "event": "push",
>         "branch": "master",
>         "commit": "436b7a6e2abaddfd35740527353e78a227ddcb2c",
>         "ref": "refs/heads/master"
>     },
>     "workspace": {
>         "root": "/drone/src",
>         "path": "/drone/src/github.com/drone/drone"
>     },
>     "vargs": {
>     }
> }
> EOF
auth: unexpected end of JSON input
```

Note: The output `auth: unexpected end of JSON input` can be optimized of course with checking required attributes ;)

**Comment by crhym3** [comment]
So, if this works, what change does this PR introduce?

**Comment by tboerger** [comment]
> I'd be glad to close this PR if you'd like to propose an alternative Makefile, @crhym3. I'm mostly just wanting to stop having to run a custom build of this plugin to use it after making some slight changes to the way auth is done.

This Makefile should be copied to the other plugins as well. And maybe we should keep the naming of the tasks as all of them are part of the drone.yml.

**Comment by tboerger** [comment]
Now it's not fixing any real issue... It's just taking the makefile from other plugins and it should add the build/buildDate output.

**Comment by crhym3** [comment]
I don't really understand why you're trying to unify every single repo. Each plugin does a different job and may have different needs. But it's up to the repo owners, of course. Personally, I wouldn't do it.

As for `build` and `buildDate`, see my comments above. It has no effect, goes to /dev/null unless you actually use it, and simply clutters otherwise clean info that's actually useful.

**Comment by bradrydzewski** [comment]
I'm going to weigh in on this one. I don't think we know the exact issue @gtaylor faced, but my guess is that he was building the binary for the wrong architecture (not setting `CGO`, `GOOS` and `GOARCH`) resulting in a busted Docker image. Using the Makefile in the pull request likely helped fixed the issue as it adds these variables.

The problem with setting `GOOS` and `GOARCH` at the top of the Makefile (per this PR) is that it impacts unit testing when I'm writing code on my Macbook. I think this has gone unnoticed thus far because many plugins don't have unit tests, but we can expect that to change. In addition, when `CGO` is set it also slows down my builds because Go has to recompile a bunch of stuff in the stdlib.

I'm therefore going to propose we do not merge this PR and close because a) we haven't identified the root cause of Gregs issue and b) we could improve upon the existing Makefile that is being added to all of the other repositories.

I think it is fine to have a standard Makefile that gets copied into new plugins, but at the same time I don't see a lot of value in updating existing repositories unless absolutely necessary. I'd rather see the effort put into new plugins for the community :)
